model/labels/labels_stringlabels.go

```diff
@@ -425,34 +425,50 @@ func FromStrings(ss ...string) Labels {
 // TODO: replace with Less function - Compare is never needed.
 // TODO: just compare the underlying strings when we don't need alphanumeric sorting.
 func Compare(a, b Labels) int {
-	l := len(a.data)
-	if len(b.data) < l {
-		l = len(b.data)
-	}
-
-	ia, ib := 0, 0
-	for ia < l {
-		var aName, bName string
-		aName, ia = decodeString(a.data, ia)
-		bName, ib = decodeString(b.data, ib)
-		if aName != bName {
-			if aName < bName {
-				return -1
-			}
-			return 1
-		}
-		var aValue, bValue string
-		aValue, ia = decodeString(a.data, ia)
-		bValue, ib = decodeString(b.data, ib)
-		if aValue != bValue {
-			if aValue < bValue {
-				return -1
-			}
-			return 1
+	// Find the first byte in the string where a and b differ.
+	shorter, longer := a.data, b.data
+	if len(b.data) < len(a.data) {
+		shorter, longer = b.data, a.data
+	}
+	if len(shorter) == 0 {
+		return len(longer)
+	}
+	i := 0
+	_ = longer[len(shorter)-1] // Get compiler to do bounds-check on longer just once here.
+	for ; i < len(shorter); i++ {
+		if shorter[i] != longer[i] {
+			break
+		}
+	}
+	firstCharDifferent := i
+	if firstCharDifferent == len(shorter) {
+		if len(shorter) == len(longer) {
+			return 0
+		} else if firstCharDifferent == len(a.data) {
```

Comment:
I read what I wrote as "if the first different character is at the end of `a.data`".

I think yours requires to be thinking about longer and shorter, which I'm not at this point.

Reply:
I understand from `if firstCharDifferent == len(shorter) {` that `firstCharDifferent` actually lost its value, as it's not _first char different_ now, but a signal saying: _ all common characters are equal, pending to check whether one string is  longer than the other one._

Which now makes me think that it would be better  to just write:

```go
	if firstCharDifferent == len(shorter) {
		return compareLengths(a.data, b.data)
	}
	// ...
}
// ...

func compareLengths(a, b []string) int {
	switch {
	case len(a) == len(b):
		return 0
	case len(a) < len(b):
		return -1
	case len(b) > len(a):
		return +1
	}
}
```

Reply:
Both options are fine to me, but I agree with @colega naming of `firstCharDifferent` is indeed misleading (premature - it might be just all equal characters. Perhaps keeping `i` and not introducing new var might help here?